node/router/src/handshake.rs

```diff
@@ -309,32 +309,24 @@ impl<N: Network> Router<N> {
 
     /// Ensure the peer is allowed to connect.
     fn ensure_peer_is_allowed(&self, listener_addr: SocketAddr) -> Result<()> {
-        // Ensure the peer IP is not this node.
+        // Ensure that it's not a self-connect attempt.
         if self.is_local_ip(&listener_addr) {
             bail!("Dropping connection request from '{listener_addr}' (attempted to self-connect)")
         }
-        // Ensure either the peer is trusted or `allow_external_peers` is true.
+        // Unknown peers are untrusted, so check if `allow_external_peers` is true.
         if !self.allow_external_peers() && !self.is_trusted(&listener_addr) {
             bail!("Dropping connection request from '{listener_addr}' (untrusted)")
         }
-        // Ensure the node is not already connecting to this peer.
+
         match self.peer_pool.write().entry(listener_addr) {
             Entry::Vacant(entry) => {
-                entry.insert(Peer::Connecting);
+                entry.insert(Peer::new_connecting(false));
             }
             Entry::Occupied(mut entry) => match entry.get_mut() {
-                Peer::Candidate(peer)
-                    if peer
-                        .restricted
-                        .map(|ts| ts.elapsed().as_secs() < Self::RADIO_SILENCE_IN_SECS)
-                        .unwrap_or(false) =>
-                {
-                    bail!("Dropping connection request from '{listener_addr}' (restricted)");
-                }
                 peer @ Peer::Candidate(_) => {
-                    let _ = mem::replace(peer, Peer::Connecting);
+                    let _ = mem::replace(peer, Peer::new_connecting(peer.is_trusted()));
```

Comment:
I'm curious if you could just write
```suggestion
                    *peer = Peer::new_connecting(peer.is_trusted());
```

Reply:
it _is_ in fact possible, but the merge was faster :smiley:; in any case, both are equivalent, so it can be done as a future cleanup